x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/execution/sequence/KeyToSequences.java

```diff
@@ -9,116 +9,128 @@
         this.keyToSequences = new LinkedHashMap<>();
-        this.keyToUntil = new LinkedHashMap<>();
-    }
-
-    private SequenceGroup[] groups(SequenceKey key) {
-        return keyToSequences.computeIfAbsent(key, k -> new SequenceGroup[listSize]);
     }
 
     SequenceGroup groupIfPresent(int stage, SequenceKey key) {
-        SequenceGroup[] groups = keyToSequences.get(key);
-        return groups == null ? null : groups[stage];
+        SequenceEntry sequenceEntry = keyToSequences.get(key);
+        return sequenceEntry == null ? null : sequenceEntry.groups[stage];
     }
 
     UntilGroup untilIfPresent(SequenceKey key) {
-        return keyToUntil.get(key);
+        SequenceEntry sequenceEntry = keyToSequences.get(key);
+        return sequenceEntry == null ? null : sequenceEntry.until;
     }
 
     void add(int stage, Sequence sequence) {
         SequenceKey key = sequence.key();
-        SequenceGroup[] groups = groups(key);
-        // create the group on demand
-        if (groups[stage] == null) {
-            groups[stage] = new SequenceGroup(key);
-        }
-        groups[stage].add(sequence);
+        SequenceEntry info = keyToSequences.computeIfAbsent(key, k -> new SequenceEntry(listSize));
+        info.add(stage, sequence);
     }
 
     void until(Iterable<KeyAndOrdinal> until) {
         for (KeyAndOrdinal keyAndOrdinal : until) {
             // ignore unknown keys
             SequenceKey key = keyAndOrdinal.key();
-            if (keyToSequences.containsKey(key)) {
-                UntilGroup group = keyToUntil.computeIfAbsent(key, UntilGroup::new);
-                group.add(keyAndOrdinal);
+            SequenceEntry sequenceEntry = keyToSequences.get(key);
+            if (sequenceEntry != null) {
+                sequenceEntry.until(keyAndOrdinal.ordinal);
             }
         }
     }
 
-    void remove(int stage, SequenceGroup group) {
-        SequenceKey key = group.key();
-        SequenceGroup[] groups = keyToSequences.get(key);
-        groups[stage] = null;
-        // clean-up the key if all groups are empty
-        boolean shouldRemoveKey = true;
-        for (SequenceGroup gp : groups) {
-            if (gp != null && gp.isEmpty() == false) {
-                shouldRemoveKey = false;
-                break;
-            }
-        }
-        if (shouldRemoveKey) {
-            keyToSequences.remove(key);
-        }
-    }
-
-    void dropUntil() {
-        // clean-up all candidates that occur before until
-        for (Entry<SequenceKey, UntilGroup> entry : keyToUntil.entrySet()) {
-            SequenceGroup[] groups = keyToSequences.get(entry.getKey());
-            if (groups != null) {
-                for (Ordinal o : entry.getValue()) {
-                    for (SequenceGroup group : groups) {
-                        if (group != null) {
-                            group.trimBefore(o);
-                        }
-                    }
-                }
-            }
-        }
```

Comment:
This method has been removed as it discarded valid sequences when dealing with windows that completed before until occurred.